### PR TITLE
Add clang docker file based on ubuntu disco

### LINKED ISSUE #7248: [CI] Create a clang build CI
## Abstract

Google's clang revision (close to upstream clang commit) is used by the ossfuzz builder in Google repo. See https://github.com/google/oss-fuzz/blob/48f321e93939354788fa7a0e1e41c5843b069ad5/infra/base-images/base-clang/checkout_build_install_llvm.sh#L44-L47

Since ossfuzz build failures are reported rather late, it would be nice to have a solidity CI that uses this or perhaps the latest stable clang version to build solidity.

## Motivation

@ekpyron mentioned that

> The static z3 lib in our docker image is incompatible, though, so we could either disable Z3 for that run or we'd need a separate CI docker image with a shared z3 lib for it...
Splitting into one gcc and one clang docker image might generally make sense, though, since they will probably both be smaller and therefore faster to extract...

so one approach would be to create a new clang specific docker image. See https://github.com/ethereum/solidity/issues/7253#issuecomment-521283967

## Specification

TBD

## Backwards Compatibility

NA

### PR DESCRIPTION
**Pull request by bshastry**
fixes #7248 

In draft mode because I still haven't figured out how to properly link solidity against libc++ the way, oss-fuzz builder does (CC https://github.com/google/oss-fuzz/issues/2826)

I'm not sure we should be linking against libc++ though. This image is sufficient to build solidity and run tests.

### CODE REVIEW

**Comment by bshastry** [comment]
Oops, I meant to make this draft mode but didn't actually do. Now I don't know how to switch this to draft :-)

**Comment by bshastry** [comment]
This PR is ready for review. Once it is merged, I will push the new clang docker image upstream and use it here (instead of the ubuntu 19.04 base image). The main difference being the clang docker image builds all solidity deps (including z3) via clang version 8.

https://github.com/ethereum/solidity/blob/2d601a4f236e7d6e1cb0a77a6ba593527aa7ce0b/.circleci/config.yml#L352

**Comment by bshastry** [comment]
@ekpyron Could you please remind me why we need to build z3 statically? I guess it is better to have a common policy on how to build solidity dependencies i.e., statically or shared libraries. Right now it's a bit of a mess :-)

**Comment by ekpyron** [comment]
The general policy is *always* to build against shared libraries unless there's a good reason not to.
The "good reason" for Z3 is that debian and consequently ubuntu haven't updated their Z3 packages in over a decade or so - and thus are stuck on 4.4 whereas we hard-require at least 4.6+ and 4.8+ is even better. And we don't want to pollute end-users systems with requiring to install non-stock shared libraries that might be incompatible with other software - hence static Z3. Might not seem like it, but it's in fact the cleanest way to do things ;-). Any user who wants to build from sources can just install a newer version of shared Z3 and that's fine (i.e. we don't *require* static linking against Z3, we just require a newer version and we provide that newer version for the release build statically and thus also use it for our test runs to avoid surprises).

**Comment by ekpyron** [comment]
I.e. it would be perfectly valid to have a separate docker image for the clang build that contains a 
*shared* z3 library and link against that, since we won't use the same configuration for releases anyways, so no reason to use static linking for these builds.

File: .circleci/docker/Dockerfile.clang.ubuntu1904 [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	echo "deb http://ppa.launchpad.net/ethereum/cpp-build-deps/ubuntu $dist main" >> /etc/apt/sources.list ; \
+	apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1c52189c923f6ca9 ; \
+	apt-get update; \
+	apt-get install -qqy --no-install-recommends \
+		build-essential \
+		software-properties-common \
+		cmake ninja-build \
+		clang++-8 llvm-8-dev \
+		libjsoncpp-dev \
+		libleveldb1d \
```
Comment (chriseth) [inline .circleci/docker/Dockerfile.clang.ubuntu1904:39]: Do we really need that?

**Review by chriseth — APPROVED** [review]
Fine apart from the comment.

File: .circleci/docker/Dockerfile.clang.ubuntu1904 (hunk 1 above)
Comment (bshastry) [reply .circleci/docker/Dockerfile.clang.ubuntu1904:39]: I wasn't sure so I kept it during the process of adapting the current ubuntu_19.04 dockerfile

https://github.com/ethereum/solidity/blob/54d936dda839c690f6acd3b7b7c61280aaeffa05/.circleci/docker/Dockerfile.ubuntu1904#L40

**Pull request merged**